tests/testthat/test-lazyframe-frame.R

```diff
@@ -347,3 +347,206 @@ test_that("unique works", {
+    bar = c(6, 7, 8),
+    ham = c("a", "b", "c")
+  )
+  other_df <- pl$DataFrame(
+    apple = c("x", "y", "z"),
+    ham = c("a", "b", "d")
+  )
+
+  # inner default
+  expect_query_equal(
+    .input$join(.input2, on = "ham"),
+    .input = df, .input2 = other_df,
+    pl$DataFrame(
+      foo = 1:2,
+      bar = c(6, 7),
+      ham = c("a", "b"),
+      apple = c("x", "y")
+    )
+  )
+
+  # outer
+  expect_query_equal(
+    .input$join(.input2, on = "ham", how = "full"),
+    .input = df, .input2 = other_df,
+    pl$DataFrame(
+      foo = c(1L, 2L, NA, 3L),
+      bar = c(6, 7, NA, 8),
+      ham = c("a", "b", NA, "c"),
+      apple = c("x", "y", "z", NA),
+      ham_right = c("a", "b", "d", NA)
+    )
+  )
+
+  # error on invalid 'how'
+  expect_query_error(
+    df$lazy()$join(other_df$lazy(), on = "ham", how = "foobar"),
+    "must be one of"
+  )
+  expect_query_error(
+    df$lazy()$join(other_df$lazy(), on = "ham", how = 42),
+    "must be a string or character vector"
+  )
+  # 'other' must be of same class
+  expect_query_error(
+    df$lazy()$join(other_df, on = "ham"),
+    "must be a polars lazy frame"
+  )
+})
+
+test_that("right join works", {
+  a <- pl$DataFrame(a = c(1, 2, 3), b = c(1, 2, 4))
+  b <- pl$DataFrame(a = c(1, 3), b = c(1, 3), c = c(1, 3))
+  expect_query_equal(
+    .input$join(.input2, on = "a", how = "right", coalesce = TRUE),
+    .input = a, .input2 = b,
+    pl$DataFrame(
+      b = c(1, 4),
+      a = c(1, 3),
+      b_right = c(1, 3),
+      c = c(1, 3)
+    )
+  )
+  expect_query_equal(
+    .input$join(.input2, on = "a", how = "right", coalesce = FALSE),
+    .input = a, .input2 = b,
+    pl$DataFrame(
+      a = c(1, 3),
+      b = c(1, 4),
+      a_right = c(1, 3),
+      b_right = c(1, 3),
+      c = c(1, 3)
+    )
+  )
+})
+
+test_that("semi and anti join", {
+  df_a <- pl$DataFrame(key = 1:3, payload = c("f", "i", NA))
+  df_b <- pl$DataFrame(key = c(3L, 4L, 5L, NA))
+
+  expect_query_equal(
+    .input$join(.input2, on = "key", how = "anti"),
+    .input = df_a, .input2 = df_b,
+    pl$DataFrame(key = 1:2, payload = c("f", "i"))
+  )
+  expect_query_equal(
+    .input$join(.input2, on = "key", how = "semi"),
+    .input = df_a, .input2 = df_b,
+    pl$DataFrame(key = 3L, payload = NA_character_)
+  )
+
+  df_a <- pl$DataFrame(a = c(1:3, 1L), b = c("a", "b", "c", "a"), payload = c(10L, 20L, 30L, 40L))
+  df_b <- pl$DataFrame(a = c(3L, 3L, 4L, 5L), b = c("c", "c", "d", "e"))
+
+  expect_query_equal(
+    .input$join(.input2, on = c("a", "b"), how = "anti"),
+    .input = df_a, .input2 = df_b,
+    pl$DataFrame(a = c(1:2, 1L), b = c("a", "b", "a"), payload = c(10L, 20L, 40L))
+  )
+  expect_query_equal(
+    .input$join(.input2, on = c("a", "b"), how = "semi"),
+    .input = df_a, .input2 = df_b,
+    pl$DataFrame(a = 3L, b = "c", payload = 30L)
+  )
+})
+
+# TODO-REWRITE: panics
+# test_that("cross join", {
+#   dat <- pl$DataFrame(x = letters[1:3])
+#   dat2 <- pl$DataFrame(y = 1:4)
```

Comment:
Very confused about why this panics. Reprex:
```r
dat <- pl$LazyFrame(x = "a")
dat2 <- pl$LazyFrame(x = "a")
dat$join(dat2, how = "cross")$collect()
```
The join doesn't error in itself, collecting does.